# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [71]:
import pandas as pd

In [72]:
taxi_data = pd.read_csv('https://data.cityofnewyork.us/resource/2upf-qytp.csv', nrows=5)
taxi_data

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-04-06T18:23:22.000  2019-04-06T18:40:20.000   
1         2  2019-04-06T18:47:35.000  2019-04-06T18:56:52.000   
2         1  2019-04-06T18:15:27.000  2019-04-06T19:16:09.000   
3         2  2019-04-06T18:18:29.000  2019-04-06T18:30:32.000   
4         2  2019-04-06T18:46:35.000  2019-04-06T18:49:19.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           2.60           1                  N   
1                2           1.70           1                  N   
2                2          17.10           2                  N   
3                1           1.40           1                  N   
4                1           0.36           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           161           238             1         13.0    0.0      0.5   
1            43           140             2          8.5    0.0      0.5   
2           234           132             2         52.0    2.5      0.5   
3            68           158             1          9.0    0.0      0.5   
4           249            90             1          4.0    0.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        0.00          0.00                    0.3         16.30   
1        0.00          0.00                    0.3         11.80   
2        0.00          6.12                    0.3         61.42   
3        1.00          0.00                    0.3         13.30   
4        1.46          0.00                    0.3          8.76   

   congestion_surcharge  
0                   2.5  
1                   2.5  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [73]:
taxi_data = pd.read_csv('https://data.cityofnewyork.us/resource/2upf-qytp.csv')
taxi_data.shape

(1000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [74]:
taxi_data[['fare_amount','tip_amount','tolls_amount','total_amount']].describe()

fare_amount   tip_amount  tolls_amount  total_amount
count  1000.000000  1000.000000   1000.000000   1000.000000
mean     13.167820     2.099180      0.370770     18.762920
std      13.528199     2.771258      1.646696     16.511177
min      -7.000000     0.000000      0.000000    -10.300000
25%       6.500000     0.000000      0.000000     10.800000
50%       9.500000     1.760000      0.000000     14.300000
75%      14.500000     2.760000      0.000000     19.800000
max     202.500000    24.540000     17.450000    229.420000

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [75]:
taxi_data[taxi_data.trip_distance==max(taxi_data.trip_distance)][['fare_amount','tip_amount','tolls_amount','total_amount']]



fare_amount  tip_amount  tolls_amount  total_amount
263        202.5        20.0          6.12        229.42

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.